tools/common/DerivedLinkContext.cs

```diff
@@ -65,5 +60,67 @@ public DerivedLinkContext (Pipeline pipeline, AssemblyResolver resolver)
 		{
 			UserAction = AssemblyAction.Link;
 		}
+
+		public Dictionary<IMetadataTokenProvider, object> GetAllCustomAttributes (string storage_name)
+		{
+			return Annotations?.GetCustomAnnotations (storage_name);
+		}
+
+		public List<ICustomAttribute> GetCustomAttributes (ICustomAttributeProvider provider, string storage_name)
+		{
+			var annotations = Annotations?.GetCustomAnnotations (storage_name);
+			object storage = null;
+			if (annotations?.TryGetValue (provider, out storage) != true)
+				return null;
+			return (List<ICustomAttribute>) storage;
+		}
+
+		// Stores custom attributes in the link context, so that the attribute can be retrieved and
+		// inspected even if it's linked away.
+		public void StoreCustomAttribute (ICustomAttributeProvider provider, CustomAttribute attribute, string storage_name)
+		{
+			var dict = Annotations.GetCustomAnnotations (storage_name);
```

Comment:
Why does this not need a ? while Get does?

```Annotations.GetCustomAnnotations```

Reply:
Just defensive programming on my side: for the `Get` it's clear that if we don't have `Annotations`, we won't have attributes either (so it makes sense to check for null, and return nothing in that case). However, what should happen if there are no `Annotations` when we're trying to store an attribute? IMHO the sanest thing is to throw an exception, since it shouldn't happen in the first place.